xds/src/main/java/io/grpc/xds/Bootstrapper.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import com.google.auth.oauth2.ComputeEngineCredentials;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.util.JsonFormat;
+import io.envoyproxy.envoy.api.v2.core.ApiConfigSource;
+import io.envoyproxy.envoy.api.v2.core.ApiConfigSource.ApiType;
+import io.envoyproxy.envoy.api.v2.core.Node;
+import io.grpc.CallCredentials;
+import io.grpc.auth.MoreCallCredentials;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * Loads configuration information to bootstrap xDS load balancer.
+ */
+@Immutable
+abstract class Bootstrapper {
+
+  private static final String BOOTSTRAP_PATH_SYS_ENV_VAR = "GRPC_XDS_BOOTSTRAP";
```

Comment:
This is not an interface. It defines a blueprint instead of a contract. The class is made abstract for user's testing purposes.

Reply:
Seems you are testing more of the contract not anything blueprint.

Reply:
It tests `Bootstrapper`'s behavior, aka, fed with some data and verify it return desired outcome. What do you mean by testing contract?

Reply:
Seems you are actually just testing the real impl class `@VisibleForTesting class FileBasedBootstrapper`. I don't see Bootstrapper is abstract out for testing purpose.

Reply:
I mean this class is abstract for user's testing purpose, for classes that uses `Bootstrapper`.

Reply:
Then isn't the user testing the contract?

Reply:
What do you mean by testing the contract?

Reply:
What do you mean by blueprint?
>It defines a blueprint instead of a contract.

Reply:
An abstract class is a blueprint for concrete instances of classes extending it, while an interface only defines a contract (aka, how objects behave) for instances of classes that implement it. Here `Bootstapper` is a blueprint for `FileBasedBootstrapper` or `FakeBootstrapper`. In general, abstract classes use noun (e.g. `AbstractMap`) for names while  interfaces use adjectives (e.g., `Closable`, `Serializable`).

Reply:
As far as I know interfaces also widely use nouns for names like `Map`, `List`, `Comparator`, `InputStream`... The `Bootstapper` doesn't hold any state or have nonabstract methods, so I don't see why you want it to be a blueprint not contract. Actually I don't see user can test anything "blueprint" from it, user can only test the "contract (aka, how objects behave)". Otherwise user just tests the default impl instance.

Reply:
If you want `Bootstapper` a blueprint, I think you might make it without abstract method, and implement a constructor
```
Bootstrapper(Bootstrap bootstrapConfig)
```

And move 
```
private static Bootstrap readConfig() 
```
to `FileBasedBootstrapper`, it may more belong to `FileBasedBootstrapper` than `Bootstrapper`

Reply:
Why do I need to test `Bootstrapper`? `Bootstrapper` is abstract for users to mock/fake. My test is for the real implementation here.